File: lib/src/features/matchmaking/domain/entities/match.dart
class Participant {
  final String id;
  final String name;
  final String? photoUrl;

  const Participant({required this.id, required this.name, this.photoUrl});
}

class MatchmakingMatch {
  final String id;
  final String title;
  final String pitchId;
  final String creatorId;
  final String date;
  final String startTime;
  final String endTime;
  final int maxPlayers;
  final List<String> participantIds;
  final List<Participant> participants;
  final List<Participant> teamA;
  final List<Participant> teamB;
  final String? winner;
  final String? cancellationReason;
  final double pricePerPlayer;
  final String skillLevel;
  final String status;
  final String? pitchName;
  final String? pitchImage;
  final String sportType;

  const MatchmakingMatch({
    required this.id,
    required this.title,
    required this.pitchId,
    required this.creatorId,
    required this.date,
    required this.startTime,
    required this.endTime,
    required this.maxPlayers,
    required this.participantIds,
    this.participants = const [],
    this.teamA = const [],
    this.teamB = const [],
    this.winner,
    this.cancellationReason,
    required this.pricePerPlayer,
    required this.skillLevel,
    required this.status,
    this.pitchName,
    this.pitchImage,
    required this.sportType,
  });

  bool get isFull => participantIds.length >= maxPlayers;
  int get slotsLeft => maxPlayers - participantIds.length;
}

File: e7gz_backend/src/models/Match.js
import mongoose from 'mongoose';

const matchSchema = new mongoose.Schema(
  {
    title: {
      type: String,
      required: [true, 'Match title is required'],
      trim: true,
    },
    pitchId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Pitch',
      required: true,
    },
    creatorId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    date: {
      type: String, // 'YYYY-MM-DD'
      required: true,
    },
    startTime: {
      type: String, // 'HH:mm'
      required: true,
    },
    endTime: {
      type: String, // 'HH:mm'
      required: true,
    },
    maxPlayers: {
      type: Number,
      required: true,
      min: 2,
    },
    participantIds: [
      {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
      },
    ],
    pricePerPlayer: {
      type: Number,
      required: true,
    },
    skillLevel: {
      type: String,
      enum: ['beginner', 'intermediate', 'advanced', 'all'],
      default: 'all',
    },
    status: {
      type: String,
      enum: ['open', 'full', 'completed', 'cancelled'],
      default: 'open',
    },
    bookingId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Booking',
    },
  },
  { timestamps: true }
);

// Virtual to check if match is full
matchSchema.virtual('isFull').get(function () {
  return this.participantIds.length >= this.maxPlayers;
});

// Auto-update status when participant joins
matchSchema.pre('save', function (next) {
  if (this.participantIds.length >= this.maxPlayers) {
    this.status = 'full';
  } else if (this.status === 'full' && this.participantIds.length < this.maxPlayers) {
    this.status = 'open';
  }
  next();
});

export default mongoose.model('Match', matchSchema);
